src/python/pants/rules/core/list_target_types.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Type
+
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem, LineOriented
+from pants.engine.rules import UnionMembership, goal_rule
+from pants.engine.target import BoolField, Field, RegisteredTargetTypes, Target
+from pants.util.objects import get_first_line_of_docstring
+
+
+class TargetTypesOptions(LineOriented, GoalSubsystem):
+    """List all the registered target types, including custom plugin types."""
+
+    # TODO: consider renaming this to to `target-types` or even a more general goal like
+    #  `symbols --type=targets`.
+    name = "targets2"
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--details",
+            type=str,
+            metavar="target_type",
+            help="List all of the target type's registered fields.",
+        )
+
+
+class TargetTypes(Goal):
+    subsystem_cls = TargetTypesOptions
+
+
+def abbreviated_target_information(target_type: Type[Target], console: Console) -> str:
+    """Return a single line description of the target type."""
+    formatted_alias = console.cyan(f"{target_type.alias:>30}:")
+    description = get_first_line_of_docstring(target_type) or "<no description>"
+    return f"{formatted_alias} {description}"
+
+
+def verbose_target_information(
+    target_type: Type[Target], console: Console, union_membership: UnionMembership
+) -> str:
```

Comment:
Typed fields make life so nice. Compare this to https://github.com/pantsbuild/pants/blob/master/src/python/pants/help/build_dictionary_info_extracter.py.